src/main/scala/fi/oph/koski/koodisto/KoodistoViitePalvelu.scala

```diff
@@ -40,15 +40,23 @@ case class KoodistoViitePalvelu(koodistoPalvelu: KoodistoPalvelu)(implicit cache
   def validate(input: Koodistokoodiviite): Option[Koodistokoodiviite] = {
     val koodistoViite = toKoodistoViiteOptional(input)
 
-    val viite = koodistoViite.flatMap(getKoodistoKoodiViitteet(_).find(_.koodiarvo == input.koodiarvo))
+    val viite = koodistoViite.flatMap { v =>
+      val koodiviitteet = getKoodistoKoodiViitteet(v)
+      koodiviitteet.find(k => k.koodiarvo == input.koodiarvo && k.koodistoVersio == input.koodistoVersio).orElse(
+        koodiviitteet.find(k => k.koodiarvo == input.koodiarvo)
```

Comment:
Ymmärrän, että vanhan toteutuksen muuttamista ei välttämättä kannata tehdä, mutta jäin tässä pohtimaan, että jos pyydetään eksplisiittisellä input.koodistoVersio:lla, jota ei löydy, tai kokonaan ilman versiota, niin onko aina oikein palauttaa viite mahdollisesti vanhemman tai uudemman koodistoversion dataan?

Varmaan hyvä jotain on palauttaa, mutta ehkä olisi loogisinta palauttaa myös deterministisesti jollain säännöllä, esim. aina ensisijaisesti uusimpaan, joka on vanhempi kuin pyydetty tms.

Reply:
Mielestäni oikeissa ympäristöissä koodiviitteet-listassa ei tässä kohtaa ole yhdellä koodiarvolla kuin yksi alkio. Tässä `getKoodistoKoodiViitteet` hakee cachesta tai koodisto-palvelusta koodiston tietyn version kaikki koodiarvot, ja ainakin APIa kokeilemalla näyttää että duplikaatteja ei tule. Luulen että sen takia alkuperäinen toteutus ei ole ollut tuon eksplisiittisempi.

Kuitenkin lokaalisti olen nyt lisännyt saman koodiarvon "duplikaattina" samaan koodistoon eri versiolla, kun sille ei vaikuta olevan mitään estettäkään, jotta saan tehdyt muutokset testattua lokaalisti. Näillä KoodistoViitePalvelun muutoksilla sitten onnistuu lokaalisti löytämään eri versiot samasta koodiarvosta.

Parempi olisi jos Kosken mock-koodistot olisi oikeasti versioitu, mutta sille ei vaikuta olevan nykyisellään tukea ja tuntui tätä tehdessä aika isolta hommalta lisätä tuki tuollaiseen.